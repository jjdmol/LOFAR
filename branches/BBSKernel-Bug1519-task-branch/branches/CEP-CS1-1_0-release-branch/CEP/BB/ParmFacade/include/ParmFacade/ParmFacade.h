//# ParmFacade.h: Data access the parameter database
//#
//# Copyright (C) 2006
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#ifndef LOFAR_PARMFACADE_PARMFACADE_H
#define LOFAR_PARMFACADE_PARMFACADE_H

// \file
// Data access the parameter database.

//# Includes
#include <ParmDB/ParmDB.h>
#include <casa/Containers/Record.h>
#include <vector>
#include <map>
#include <set>


namespace LOFAR {
namespace ParmDB {


// \ingroup ParmFacade
// @{

// ParmFacade is the high level interface to the Parameter Data Base.
// The current version assumes it is an AIPS++ table; with a few extra
// constructor arguments it can easily be changed to other types of
// databases.
//
// The class provides a few functions:
// <ul>
// <li> getNames returns a vector of the parameter names in the table
// <li> getRange returns a vector of 4 elements giving the boundary box
//      of the domains of the parameters.
// <li> getValues returns the values of the parameters a calculated on
//      a grid given by the caller.
// <li> getHistory returns the coefficients of parameters in a given
//      time frame. It shows how a solution converges (or diverges).
// </ul>
//
// The parameter names can be given as a pattern. This is the same as a
// file name pattern that can be given in the UNIX shells (e.g. RA:*).
// Thus it is not a full regular expression.

class ParmFacade
{
public:
  // Make a connection to the given ParmTable.
  ParmFacade (const std::string& tableName);

  // The destructor disconnects.
  ~ParmFacade();

  // Get the domain range (as startx,endx,starty,endy) of the given
  // parameters in the table.
  // This is the minimum start value and maximum end value for all parameters.
  // An empty name pattern is the same as * (all parm names).
  std::vector<double> getRange (const std::string& parmNamePattern = "") const;

  // Get parameter names in the table matching the pattern.
  // An empty name pattern is the same as * (all parm names).
  std::vector<std::string> getNames
                               (const std::string& parmNamePattern = "") const;

  // Get the parameter values for the given parameters and domain.
  // The domain is given by the start and end values, while the grid is
  // given by nx and ny.
  // The vector values in the map are in fact 2-dim arrays with axes nx and ny.
  // @{
  std::map<std::string, std::vector<double> >
    getValues (const std::string& parmNamePattern,
	       double startx, double endx, int nx,
	       double starty, double endy, int ny);

  // The Record contains a map of parameter name to Array<double>.
  casa::Record
    getValuesRec (const std::string& parmNamePattern,
		  double startx, double endx, int nx,
		  double starty, double endy, int ny);
  // @}

  // Get the parameter coefficients for the given given parameters and domain
  // which have been solved in the given time frame. Default is the
  // entire time frame.
  // The vector values in the map are in fact 2-dim arrays with shape
  // [ncoeff, nsolutions]. Because ncoeff and nsolutions are unknown,
  // for the caller, ncoeff is put in the vector as the first value.
  // The solutions are in order of timestamp.
  // A parameter can have multiple solve domains, so a domain name
  // (:domain<seqnr>) is added to the name to make it unique.
  // @{
  std::map<std::string, std::vector<double> >
    getHistory (const std::string& parmNamePattern,
		double startx, double endx,
		double starty, double endy,
		double startSolveTime=0, double endSolveTime=1e25);

  // The Record contains a map of parameter name to Array<double>.
  casa::Record
    getHistoryRec (const std::string& parmNamePattern,
		   double startx, double endx,
		   double starty, double endy,
		   double startSolveTime=0, double endSolveTime=1e25);
  // @}

  // Get the expressions and the names of their children.
  const std::map<std::string,std::set<std::string> >& getExprs() const
    { return itsExprs; }

private:
  // Process a parmexpr and add the children names to the vector.
  // It is done recursively, so a child can be an expression.
  void processExpr (const std::map<std::string,ParmValueSet>& defValues, 
		    const std::string& expr, std::set<std::string>& ch);

  // Copy the values from a record to a map.
  std::map<std::string, std::vector<double> >
    record2Map (const casa::Record& rec) const;

  // Compare function for sorting ParmValues in time order.
  static bool pvLess (const ParmValue& left, const ParmValue& right);


  ParmDB itsPDB;
  std::map<std::string,std::set<std::string> > itsExprs;
};

// @}

} // namespace ParmDB
} // namespace LOFAR

#endif
