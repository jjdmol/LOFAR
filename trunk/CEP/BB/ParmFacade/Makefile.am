SUBDIRS=src test include

DISTCHECK_CONFIGURE_FLAGS=\
      --with-common=$(prefix)

EXTRA_DIST = \
      Makefile.common \
      ParmAcc.spec \
      autoconf_share/compiletool

include $(top_srcdir)/Makefile.common
