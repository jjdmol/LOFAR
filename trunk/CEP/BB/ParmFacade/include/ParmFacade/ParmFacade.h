//# ParmFacade.h: Data access the parameter database
//#
//# Copyright (C) 2006
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#ifndef LOFAR_PARMFACADE_PARMFACADE_H
#define LOFAR_PARMFACADE_PARMFACADE_H

// \file
// Data access the parameter database.

//# Includes
#include <ParmDB/ParmDB.h>
#include <casa/Containers/Record.h>
#include <Common/lofar_vector.h>
#include <Common/lofar_map.h>
#include <Common/lofar_set.h>
#include <Common/lofar_string.h>


namespace LOFAR { namespace BBS {


  // \ingroup ParmFacade
  // @{

  // ParmFacade is the high level interface to the Parameter Data Base.
  // The current version assumes it is an AIPS++ table; with a few extra
  // constructor arguments it can easily be changed to other types of
  // databases.
  //
  // The class provides a few functions:
  // <ul>
  // <li> getNames returns a vector of the parameter names in the table
  // <li> getRange returns a vector of 4 elements giving the boundary box
  //    of the domains of the parameters.
  // <li> getValues returns the values of the parameters a calculated on
  //      a grid given by the caller.
  // </ul>
  //
  // The parameter names can be given as a pattern. This is the same as a
  // file name pattern that can be given in the UNIX shells (e.g. RA:*).
  // Thus it is not a full regular expression.

  class ParmFacade
  {
  public:
    // Make a connection to the given ParmTable.
    ParmFacade (const string& tableName);

    // The destructor disconnects.
    ~ParmFacade();

    // Get the domain range (as startx,endx,starty,endy) of the given
    // parameters in the table.
    // This is the minimum start value and maximum end value for all parameters.
    // An empty name pattern is the same as * (all parm names).
    vector<double> getRange (const string& parmNamePattern = "") const;

    // Get parameter names in the table matching the pattern.
    // An empty name pattern is the same as * (all parm names).
    vector<string> getNames (const string& parmNamePattern = "") const;

    // Get the parameter values for the given parameters and domain.
    // The domain is given by the start and end values, while the grid is
    // given by nx and ny.
    // The vector values in the map are in fact 2-dim arrays with axes nx and ny.
    // @{
    map<string, vector<double> > getValues (const string& parmNamePattern,
                                            double startx, double endx, int nx,
                                            double starty, double endy, int ny);

    // The Record contains a map of parameter name to Array<double>.
    casa::Record getValuesRec (const string& parmNamePattern,
                               double startx, double endx, int nx,
                               double starty, double endy, int ny);
    // @}

  private:
    // Copy the values from a record to a map.
    map<string, vector<double> > record2Map (const casa::Record& rec) const;


    //# Data members
    ParmDB itsPDB;
  };

  // @}

}} // namespaces

#endif
