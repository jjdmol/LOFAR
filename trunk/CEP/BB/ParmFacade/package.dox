// \ingroup BB
// \defgroup ParmFacade ParmFacade High level interface to Parameter Data Base
