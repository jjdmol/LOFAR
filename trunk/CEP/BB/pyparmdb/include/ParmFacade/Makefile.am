INSTHDRS = \
ParmFacade.h

NOINSTHDRS =

TCCHDRS =

pkginclude_HEADERS = Package__Version.h  $(INSTHDRS) $(TCCHDRS)

noinst_HEADERS = $(NOINSTHDRS)

DOCHDRS = $(INSTHDRS) $(NOINSTHDRS)

include $(top_srcdir)/Makefile.common
