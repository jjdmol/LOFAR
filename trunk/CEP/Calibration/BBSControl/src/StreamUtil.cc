//# StreamUtil.cc: Specializations of the <Common/StreamUtil.cc>.
//#
//# Copyright (C) 2002-2008
//# ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <BBSControl/StreamUtil.h>

namespace LOFAR
{
  template<>
  void writeVector (ostream& os, const vector<string>& vec,
                    const char* separator,
                    const char* prefix, const char* postfix)
  {
    os << prefix;
    for (unsigned int i = 0; i < vec.size(); i++) {
      if (i > 0) os << separator;
      os << '"';
      const string& s = vec[i];
      for (unsigned int j = 0; j < s.size(); ++j) {
        if (s[j] == '"') os << '\\';
        os << s[j];
      }
      os << '"';
    }
    os << postfix;
  }

} // namespace LOFAR
