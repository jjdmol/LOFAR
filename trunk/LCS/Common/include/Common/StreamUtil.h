//# StreamUtil.h: useful stream manipulation methods.
//#
//# Copyright (C) 2002-2003
//# ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id$

#ifndef LOFAR_COMMON_STREAMUTIL_H
#define LOFAR_COMMON_STREAMUTIL_H

// \file
// Useful stream manipulation methods.

//# Includes
#include <utility>
#include <Common/LofarTypes.h>
#include <Common/lofar_string.h>
#include <Common/lofar_vector.h>
#include <Common/lofar_map.h>
#include <Common/lofar_iostream.h>

namespace LOFAR
{
  // Handles indentation of text lines. Every time an Indent object is
  // constructed, the static member \c level is incremented by one, and every
  // time an Indent object is destructed \c level is decremented by one. To
  // increment the amount of indentation you simply create an Indent
  // object. When this object goes out of scope, the amount of indentation is
  // automagically decremented.
  class Indent {
  public:
    // Constructor. Increments indentation level.
    Indent() { lvl++; }
    // Destructor. Decrements indentation level.
    ~Indent() { lvl--; }
    // Return the amount of indentation.
    static uint level() { return lvl; }
    // Return the token to be printed per indentation level
    static const string& token() { return tok; }
  private:
    // Indentation level.
    static uint lvl;
    // Token to be printed per indentation level.
    static const string tok;
  };

  // Print an indentation that depends on the number of Indent objects
  // currently in existence.
  inline ostream& indent(ostream& os) 
  {
    for (uint i = 0; i < Indent::level(); ++i) {
      os << Indent::token();
    }
    return os;
  }


  // Declare the functions.
  // Write a std::pair.
  template <typename T, typename U>
  ostream& operator<< (ostream& os, const std::pair<T,U>& p);
  // Write any container to the given ostream.
  template<typename ITER>
  void print (ostream& os, ITER begin, ITER end, 
              const char* separator=",",
              const char* prefix="[", const char* postfix="]");
  // Write a vector to an ostream with a given separator, prefix and postfix.
  template<class T>
  void writeVector (std::ostream& os, const std::vector<T>& vec,
		    const char* separator=",",
		    const char* prefix="[", const char* postfix="]");
  // Print the contents of a vector enclosed in square brackets, using a comma
  // as separator.
  // \note operator<<() must be defined for type \c T.
  template<typename T>
  ostream& operator<<(ostream& os, const vector<T>& v);
  // Print the contents of a map enclosed in braces, using a comma
  // as separator.
  // \note operator<<() must be defined for type \c T.
  template<typename T, typename U>
  ostream& operator<<(ostream& os, const map<T,U>& m);


  // Write a std::pair.
  template <typename T, typename U>
  inline ostream& operator<< (ostream& os, const std::pair<T,U>& p)
  {
    os << '<' << p.first << ',' << p.second << '>';
    return os;
  }

  // Write any container to the given ostream.
  template<typename ITER>
  inline void print (ostream& os, ITER begin, ITER end, 
                     const char* separator,
                     const char* prefix, const char* postfix)
  {
    os << prefix;
    if (begin != end) {
      os << *begin;
      ++begin;
    }
    for (; begin!=end; ++begin) {
      os << separator << *begin;
    }
    os << postfix;
  }


  // Write a vector to an ostream with a given separator, prefix and postfix.
  template<class T>
  inline void writeVector (std::ostream& os, const std::vector<T>& vec,
                           const char* separator,
                           const char* prefix, const char* postfix)
  {
    print (os, vec.begin(), vec.end(), separator, prefix, postfix);
  }


  // Print the contents of a vector enclosed in square brackets, using a comma
  // as separator.
  // \note operator<<() must be defined for type \c T.
  template<typename T>
  inline ostream& operator<<(ostream& os, const vector<T>& v)
  {
    writeVector(os, v, ",", "[", "]");
    return os;
  }


  // Print the contents of a map enclosed in braces, using a comma
  // as separator.
  // \note operator<<() must be defined for type \c T.
  template<typename T, typename U>
  inline ostream& operator<<(ostream& os, const map<T,U>& m)
  {
    print (os, m.begin(), m.end(), ", ", "{", "}");
    return os;
  }


} // namespace LOFAR

#endif
