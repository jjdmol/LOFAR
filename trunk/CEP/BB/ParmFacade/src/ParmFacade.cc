//# ParmFacade.cc: Object access the parameter database
//#
//# Copyright (C) 2006
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#include <ParmFacade/ParmFacade.h>
#include <BBSKernel/MNS/MeqRequest.h>
#include <BBSKernel/MNS/MeqParmFunklet.h>
#include <BBSKernel/MNS/MeqResult.h>
#include <BBSKernel/MNS/MeqMatrix.h>
#include <Common/LofarLogger.h>
#include <Common/StringUtil.h>
#include <casa/Utilities/Regex.h>

using namespace std;
using namespace casa;
using namespace LOFAR::BBS;

// Create tParmFacade.in_mep with parmdb using:
//   create tablename='tParmFacade.in_mep'
//   add parm1 domain=[1,5,4,10],values=2
//   add parm2 domain=[1,5,4,10],values=[2,0.1],nx=2
//   add parm3 type='expression',expression='parm1*parm2'

namespace LOFAR {
namespace ParmDB {

ParmFacade::ParmFacade (const string& tableName)
    : itsPDB(ParmDBMeta("aips", tableName))
{
    map<string,ParmValueSet> defValues;
    itsPDB.getDefValues (defValues, "*");
    for (map<string,ParmValueSet>::const_iterator iter = defValues.begin();
    iter != defValues.end();
    iter++) {
    const ParmValueRep& pval = iter->second.getValues()[0].rep();
    if (pval.itsType == "parmexpr") {
    itsExprs[iter->first] = set<string>();
    // Add the expression to the map.
    map<string,set<string> >::iterator ch = itsExprs.find(iter->first);
    // Add all children to the set.
    processExpr (defValues, pval.itsExpr, ch->second);
    }
    }
}

ParmFacade::~ParmFacade()
{}

void ParmFacade::processExpr (const map<string,ParmValueSet>& defValues, 
                const string& expr, set<string>& ch)
{
    String str(expr);
    // Replace function names by a blank..
    str.gsub (Regex("[a-zA-Z_][a-zA-Z0-9_]* *\\("), " ");
    // Replace parentheses, commas, and operators by a blank.
    str.gsub (Regex("[(),*/+^-]") , " ");
    // Replace multiple blanks by a single one.
    str.gsub (Regex("  *") , " ");
    // Now split the string and append to the children set.
    vector<string> nms = StringUtil::split(str, ' ');
    for (vector<string>::const_iterator iter=nms.begin();
    iter != nms.end();
    iter++) {
    if (! iter->empty()) {
    // If the name is an expression, process it recursively.
    bool isExpr = false;
    map<string,ParmValueSet>::const_iterator def = defValues.find (*iter);
    if (def != defValues.end()) {
    const ParmValueRep& pval = def->second.getValues()[0].rep();
    if (pval.itsType == "parmexpr") {
        isExpr = true;
        processExpr (defValues, pval.itsExpr, ch);
    }
    }
    if (!isExpr) {
    // No expression, so add parm name.
    ch.insert (*iter);
    }
    }
    }
}

vector<double> ParmFacade::getRange (const string& parmNamePattern) const
{
    string pp = parmNamePattern;
    if (pp.empty()) {
    pp = "*";
    }
    ParmDomain dom = itsPDB.getRange (pp);
    // If no domain found, see if the name matches an expression.
    if (dom.getStart()[0] == 0  &&  dom.getEnd()[0] == 1  &&
    dom.getStart()[1] == 0  &&  dom.getEnd()[1] == 1) {
    map<string,set<string> >::const_iterator iter =
    itsExprs.find (parmNamePattern);
    if (iter == itsExprs.end()) {
    // Not found; see if the name is part of an expression name.
    String pname(parmNamePattern);
    for (iter=itsExprs.begin(); iter!=itsExprs.end(); iter++) {
    if (pname.matches (Regex(iter->first + ".*"))) {
        break;
    }
    }
    }
    // A match has been found, so get the range of the expression's children.
    if (iter != itsExprs.end()) {
    // Copy the set to a vector.
    vector<string> nams(iter->second.begin(), iter->second.end());
    dom = itsPDB.getRange (nams);
    }
    }
    vector<double> res(4);
    res[0] = dom.getStart()[0];
    res[1] = dom.getEnd()[0];
    res[2] = dom.getStart()[1];
    res[3] = dom.getEnd()[1];
    return res;
}

// Get all parameter names in the table.
vector<string> ParmFacade::getNames (const string& parmNamePattern) const
{
    string pp = parmNamePattern;
    if (pp.empty()) {
    pp = "*";
    }
    return itsPDB.getAllNames(pp);
}

// Get the parameter values for the given parameters and domain.
map<string, vector<double> >
    ParmFacade::getValues (const string& parmNamePattern,
            double startx, double endx, int nx,
            double starty, double endy, int ny)
{
    // Get all parm names.
    vector<string> names = getNames (parmNamePattern);
    // The output is returned in a map.
    map<string,vector<double> > out;
    MeqParmGroup parmGroup;
    map<string,MeqExpr> exprs;
    // Loop through all parameters.
    for (vector<string>::const_iterator iter=names.begin();
    iter != names.end();
    iter++) {
    // Create an expression object.
    exprs.insert (std::make_pair (*iter,
                    MeqParmFunklet::create (*iter,
                                parmGroup,
                                &itsPDB)));
    }
    // Get the parm values of parameters.
    // All parameters are needed to deal with possible parmexpressions.
    map<string,ParmValueSet> parmValues;
    itsPDB.getValues (parmValues, vector<string>(),
            ParmDomain(startx, endx, starty, endy));
    // Make a request object for the given grid.
    MeqDomain domain(startx, endx, starty, endy);
    MeqRequest req(domain, nx, ny);
    // Initialize all parameters.
    //const vector<MeqParm*>& parmList = parmGroup.getParms();
    //for (vector<MeqParm*>::const_iterator iter = parmList.begin();
    // iter != parmList.end();
    // ++iter) {
    //  (*iter)->fillFunklets (parmValues, domain);
    //}
    for(MeqParmGroup::iterator iter = parmGroup.begin(); iter != parmGroup.end(); ++iter)
    {
        (*iter).second.fillFunklets (parmValues, domain);
    }
    // Evaluate all expressions and store the result in the output map.
    for (map<string,MeqExpr>::iterator iter = exprs.begin();
    iter != exprs.end();
    ++iter) {
    // Add it to the output map and get a reference to the vector.
    out.insert (std::make_pair (iter->first, vector<double>()));
    map<string,vector<double> >::iterator outiter = out.find (iter->first);
    vector<double>& vec = outiter->second;
    // Get result and put in map.
    MeqResult resVal = iter->second.getResult (req);
    const MeqMatrix& mat = resVal.getValue();
    const double* ptr = mat.doubleStorage();
    // Store the result in the vector.
    // If the result is a scalar, expand it.
    if (mat.nelements() == 1) {
    vec.assign (nx*ny, *ptr);
    } else {
    ASSERT (nx*ny == mat.nelements());
    vec.assign (ptr, ptr+nx*ny);
    }
    }
    return out;
}

// Get the coefficients of parameters of type 'history' matching the parmNamePattern
// and the specified domain. This is a temporary hack to facillitate viewing the solve
// history in the SAS GUI.
map<string, vector<double> >
    ParmFacade::getHistory (const string& parmNamePattern,
                double startx, double endx,
                double starty, double endy,
                double startSolveTime, double endSolveTime)
{
    // Get ParmValues of all matching parameters (all timestamps).
    map<string, ParmValueSet> parmValues;
    itsPDB.getValues(parmValues, parmNamePattern, ParmDomain(startx, endx, starty, endy), -1);

    // The output is returned in a map.
    map<string, vector<double> > out;

    // Loop over all parameters. A parameter can have multiple ParmValues, one for each
    // timestamp. Parameters of type 'history' have exactly one coefficient. For each
    // parameter, create an array of coefficients (one coefficient per timestamp),
    // sorted on timestamp.
    for (map<string,ParmValueSet>::iterator it = parmValues.begin(); it != parmValues.end(); ++it)
    {
        vector<double> coeff;
        vector<ParmValue>& values = it->second.getValues();
        
        // Sort ParmValues on timestamp.
        sort(values.begin(), values.end(), &pvLess);
        
        for (vector<ParmValue>::const_iterator value_it = values.begin(); value_it != values.end(); ++value_it)
        {
            // Only take ParmValue's of type 'history' with a timestamp within the specified time range.
            if(value_it->rep().itsType == "history" && 
               value_it->rep().itsTimeStamp >= startSolveTime &&
               value_it->rep().itsTimeStamp <= endSolveTime)
            {
                ASSERTSTR(value_it->rep().itsCoeff.size() == 1, "parms of type 'history' should have exactly one coefficient.");
                coeff.push_back(value_it->rep().itsCoeff[0]);
            }
        }
        out[it->first] = coeff;
    }        
    
    return out;
}

bool ParmFacade::pvLess (const ParmValue& left, const ParmValue& right)
{
    return (left.rep().itsTimeStamp < right.rep().itsTimeStamp);
}

} // namespace ParmDB
} // namespace LOFAR
