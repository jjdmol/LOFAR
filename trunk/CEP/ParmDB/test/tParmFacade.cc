//# tParmFacade.cc: test program for class ParmFacade
//#
//# Copyright (C) 2006
//# ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id$

#include <lofar_config.h>
#include <ParmDB/ParmFacade.h>
#include <Common/StreamUtil.h>
#include <casa/Containers/Record.h>
#include <casa/Arrays/ArrayIO.h>
#include <iostream>
#include <sstream>

using namespace LOFAR;
using namespace BBS;
using namespace casa;
using namespace std;

// tParmFacade.run creates the pdb with parmdb.
// So it must match the output expected here.

void showRec (const Record& rec)
{
  cout << ">start<" << endl;
  for (uint i=0; i<rec.nfields(); ++i) {
    Record rc = rec.subRecord (i);
    cout << rec.name(i) << ' ' << rc.asArrayDouble("values") << endl;
    cout << " freqs: " << rc.asArrayDouble("freqs") << endl;
    cout << " times: " << rc.asArrayDouble("times") << endl;
    cout << " freqwidths: " << rc.asArrayDouble("freqwidths") << endl;
    cout << " timewidths: " << rc.asArrayDouble("timewidths") << endl;
    cout << endl;
  }
  cout << ">end<" << endl;
}

void showValues (ParmFacade& acc, const string& pattern, int nf, int nt)
{
  // Now get the values of all parameters.
  vector<string> names = acc.getNames(pattern);
  cout << ">start<" << endl;
  cout << "Names: " << names << endl;
  vector<double> rng = acc.getRange("*");
  cout << "Range: " << rng << endl;
  cout << ">end<" << endl;
  double fs = (rng[1] - rng[0]) / nf;
  double ts = (rng[3] - rng[2]) / nt;
  map<string,vector<double> > values = acc.getValuesMap (pattern,
                                                         rng[0], rng[1], fs,
                                                         rng[2], rng[3], ts,
                                                         true);
  cout << ">start<" << endl;
  for (map<string,vector<double> >::const_iterator iter=values.begin();
       iter != values.end();
       iter++) {
    cout <<iter->first << ' ' << iter->second << endl;
  }
  cout << endl;
  cout << ">end<" << endl;
  // Get values without giving a grid.
  // Exactly matching domain.
  showRec (acc.getValuesGrid (pattern, rng[0], rng[1], rng[2], rng[3]));
  // Much bigger domain.
  showRec (acc.getValuesGrid (pattern, 0,20,0,20));
  // No matches.
  showRec (acc.getValuesGrid (pattern, 20,40,20,40));
  // A smaller domain (so using only 2nd part for tParmFacadeDistr).
  showRec (acc.getValuesGrid (pattern, 10,16,6,20));
  // Get the coeff and errors.
  cout << ">start<" << endl;
  cout << acc.getCoeff (pattern) << endl;
  cout << ">end<" << endl;
}

int main (int argc, const char* argv[])
{
  try {
    if (argc < 2) {
      cerr << "Run as: tParmFacade parmtable [pattern nf nt]"
	   << endl;
      return 1;
    }
    string pattern ="*";
    int nf = 4;
    int nt = 2;
    if (argc > 2) {
      pattern = argv[2];
    }
    if (argc > 3) {
      istringstream istr(argv[3]);
      istr >> nf;
      if (nf <= 0) nf = 1;
    }
    if (argc > 4) {
      istringstream istr(argv[4]);
      istr >> nt;
      if (nt <= 0) nt = 1;
    }
    // Open the facade.
    ParmFacade acc(argv[1]);
    // Get values.
    showValues (acc, pattern, nf, nt);
  } catch (std::exception& x) {
    cerr << "Unexpected exception: " << x.what() << endl;
    return 1;
  }
  return 0;
}
