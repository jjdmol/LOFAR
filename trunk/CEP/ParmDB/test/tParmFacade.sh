#!/bin/sh
$lofar_sharedir/runtest.sh tParmFacade > tParmFacade.log 2>&1
