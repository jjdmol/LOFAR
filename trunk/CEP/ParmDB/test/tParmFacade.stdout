>start<
Names: [parm1,parm2]
Range: [1,4,5,10]
>end<
>start<
parm1 [2,2,2,2,2,2,2,2]
parm2 [2.0125,2.0375,2.0625,2.0875,2.0125,2.0375,2.0625,2.0875]

>end<
>start<
parm1 Axis Lengths: [1, 1]  (NB: Matrix in Row/Column order)
[2]

parm2 Axis Lengths: [1, 1]  (NB: Matrix in Row/Column order)
[2.05]

parm1;freqs [2.5]
parm1;times [7.5]
parm1;freqwidths [3]
parm1;timewidths [5]
parm2;freqs [2.5]
parm2;times [7.5]
parm2;freqwidths [3]
parm2;timewidths [5]
>end<
>start<
parm1 Axis Lengths: [1, 1]  (NB: Matrix in Row/Column order)
[2]

parm2 Axis Lengths: [1, 1]  (NB: Matrix in Row/Column order)
[2.05]

parm1;freqs [2.5]
parm1;times [7.5]
parm1;freqwidths [3]
parm1;timewidths [5]
parm2;freqs [2.5]
parm2;times [7.5]
parm2;freqwidths [3]
parm2;timewidths [5]
>end<
>start<
>end<
>start<
>end<
  parm1: {
    coeff: Double array with shape [1, 1]
      [2]
    errors: Double array with shape [1, 1]
      [-1]
    freqs: Double array with shape [1]
      [2.5]
    times: Double array with shape [1]
      [7.5]
    freqwidths: Double array with shape [1]
      [3]
    timewidths: Double array with shape [1]
      [5]
  }
  parm2: {
    coeff: Double array with shape [2, 1, 1, 1]
      [2, 0.1]
    errors: Double array with shape [2, 1, 1, 1]
      [-1, -1]
    freqs: Double array with shape [1]
      [2.5]
    times: Double array with shape [1]
      [7.5]
    freqwidths: Double array with shape [1]
      [3]
    timewidths: Double array with shape [1]
      [5]
  }


>start<
Names: [parm1,parm2]
Range: [1,4,5,10]
>end<
>start<
parm1 [2,2]
parm2 [2.05,2.05]

>end<
>start<
parm1 Axis Lengths: [1, 1]  (NB: Matrix in Row/Column order)
[2]

parm2 Axis Lengths: [1, 1]  (NB: Matrix in Row/Column order)
[2.05]

parm1;freqs [2.5]
parm1;times [7.5]
parm1;freqwidths [3]
parm1;timewidths [5]
parm2;freqs [2.5]
parm2;times [7.5]
parm2;freqwidths [3]
parm2;timewidths [5]
>end<
>start<
parm1 Axis Lengths: [1, 1]  (NB: Matrix in Row/Column order)
[2]

parm2 Axis Lengths: [1, 1]  (NB: Matrix in Row/Column order)
[2.05]

parm1;freqs [2.5]
parm1;times [7.5]
parm1;freqwidths [3]
parm1;timewidths [5]
parm2;freqs [2.5]
parm2;times [7.5]
parm2;freqwidths [3]
parm2;timewidths [5]
>end<
>start<
>end<
>start<
>end<
  parm1: {
    coeff: Double array with shape [1, 1]
      [2]
    errors: Double array with shape [1, 1]
      [-1]
    freqs: Double array with shape [1]
      [2.5]
    times: Double array with shape [1]
      [7.5]
    freqwidths: Double array with shape [1]
      [3]
    timewidths: Double array with shape [1]
      [5]
  }
  parm2: {
    coeff: Double array with shape [2, 1, 1, 1]
      [2, 0.1]
    errors: Double array with shape [2, 1, 1, 1]
      [-1, -1]
    freqs: Double array with shape [1]
      [2.5]
    times: Double array with shape [1]
      [7.5]
    freqwidths: Double array with shape [1]
      [3]
    timewidths: Double array with shape [1]
      [5]
  }

